### DISCUSSION TITLE
Using DuckDb as a SQL data source for Perspective

### DISCUSSION THREAD

**Discussion by stevesimmons** [perspective-dev/perspective#1664] [Ideas]
Has anyone tried [DuckDb](https://duckdb.org/) as a data source to drive Perspective?
That could give simple Parquet input and SQL querying, with Arrow output into the Perspective grid.

Another potentially useful package here is [node-duckdb](https://github.com/deepcrawl/node-duckdb), a TypeScript DuckDb implementation.

Thanks!

**Comment by texodus** [comment]
Sorry about the late reply @stevesimmons 

It is pretty trivial to build an app that uses DuckDB's Arrow writing support to generate data suitable to be consumed by Perspective, in Browser/Wasm, Node or Python.  You'd need to be careful what queries you generate, as Perspective cannot currently read all Arrow types.  

However, Perspective is basically already a SQL engine for streaming queries, it just doesn't use SQL-the-language for queries (yet).  DuckDB has some really nice static features itself, but I am skeptical the right solution to bridge the feature gap is to string these two query engines together end-to-end, at least as far as providing a coherent UX is concerned.

`<perspective-viewer>` is _very_ tightly coupled to its wasm data engine, and replacing this outright with DuckDB or anything else is out-of-scope for the project as it stands today.

**Reply by AndreaGernone** [reply]
4 years passed. Is this answer still valid? Has anyone tried it?

**Reply by salahfh** [reply]
Check https://perspective-dev.github.io/guide/how_to/python/virtual_server/duckdb.html

**Reply by AndreaGernone** [reply]
Thank you! I saw this earlier. Since I'm thinking about using it I wanted to know if anyone else had feedback on that too

**Comment by texodus** [comment]
No, this answer is not still valid.

Perspective now has full support for DuckDB (and others) as a [Virtual Server](https://perspective-dev.github.io/guide/explanation/virtual_servers.html). This support fully replaces Perspective's built-in engine with DuckDB, including all calculation, aggregation, pivotting and even the Expression editor becomes DuckDB expressions (with highlighting and error validation). Like with the built-in engine, `<perspective-viewer>` never serializes this data in full (and uses Arrow IPC throughout where it serializes slices) so using DuckDB as a backend is never bottle-necked by Perspective ingest (there is no ingest). Tested in Python, Rust and Browser + Node.js (WASM).

Perspective DuckDB support is fast, supports the range of DuckDB plugins, and has a much higher headroom than Perspective's built-in engine due to its OPFS support (even in WASM64 mode). It does not natively support streaming updates yet, however.

[DuckDB WASM Example](https://perspective-dev.github.io/block.html?example=duckdb) with a logger that logs the internal SQL to another Perspective Table.

[Another off-site DuckDB WASM Example](https://prospectiveco.github.io/duckdb-shell/) with the `duckdb-shell` and a `<perspective-workspace>` (that has very cool DuckLake support!)

[DuckDB Python example](https://github.com/perspective-dev/perspective/blob/master/examples/python-duckdb-virtual/server.py)

**Reply by AndreaGernone** [reply]
Thank you so much for your answer!
